Add support for user namespacing of CloudFormation resources

jeancochrane opened the pull request:
## Overview

Prompt the user for a `Slug` parameter that will be used to namespace AWS resources in CloudFormation. Make sure that `Slug` is optional, such that CloudFormation reverts to sensible defaults in its absence.

Closes https://github.com/azavea/raster-vision-aws/issues/11.

## Notes

* The [original issue](https://github.com/azavea/raster-vision-aws/issues/11) noted that namespacing the ECR repo is particularly delicate. I agree, and decided not to change it in this PR, such that the repository name is still set as a parameter value (`RepositoryName`) that the user can adjust manually. My reasoning here is that I'm anticipating in https://github.com/azavea/raster-vision-aws/issues/12 that we'll want to make pulling the latest image from Quay the default behavior, and alter the `RepositoryName`, `ImageName`, and `ImageTag` parameters to be optional advanced overrides. I think this is ultimately the most convenient behavior for the template, since building and pushing your own image adds a lot of overhead to this workflow.

## Testing instructions

- Confirm that the template syntax is valid:

```
aws cloudformation validate-template --template-body file://cloudformation/template.yml
```

- Log into the R&D AWS console
- Navigate to `CloudFormation > Create Stack`
- In the `Choose a template field`, select `Upload a template to Amazon S3` and upload the template in `cloudformation/template.yml`
- Where available, use the default parameters. Specify the following required parameters:
    - `Stack Name`: `TestingCloudFormation`
    - `Slug`: A slug of your choosing
    - `VPC`: `vpc-074e8ae6256f7c1ca` (the first one on the dropdown)
    - `Subnets`: `subnet-033d2cbe8268c3067` (the first one on the dropdown)
    - `SSH Key Name`: `raster-vision` (doesn't matter unless you want to shell into the instance)
- Accept all default options on the Options screen
- Accept `I acknowledge that AWS CloudFormation might create IAM resources with custom names` on the Review screen
- Create the stack and confirm that it builds correctly
- Inspect resources and confirm that the namespacing works
- Delete the stack and confirm that everything cleans up correctly

Commit:
Add support for user namespacing of CloudFormation resources

Prompt the user for a 'Slug' parameter that will be used to namespace
AWS resources in CloudFormation. Make sure that 'Slug' is optional,
such that CloudFormation reverts to a sensible default string in its
absence.

jeancochrane commented on cloudformation/template.yml:
```diff
@@ -75,12 +71,23 @@ Parameters:
     Default: Unknown
     Description: Environment of the project being deployed (e.g. Staging, Development, Production)
+    MinLength: 1
+    AllowedPattern: ^[a-zA-Z0-9]*$
```
If the environment name is not purely alphanumeric, the resource naming conventions won't work.

Commit:
Add ConstraintDescriptions to parameters with AllowedPatterns

CloudFormation allows you to specify ConstraintDescriptions that will
get displayed to the user in the case that an AllowedPattern regex
match fails. Update parameters that use AllowedPattern to provide
explicit ConstraintDescriptions.

hectcastro commented:
Swapping myself out as a reviewer in favor of @rbreslow.

rbreslow commented on cloudformation/template.yml:
```diff
@@ -75,12 +71,25 @@ Parameters:
     Type: String
     Default: Raster Vision CloudFormation
-    Description: Name of the project associated with the Batch job
+    Description: Name of the project to associate with new resources
```
You might be able to make this more clear by saying "Name of the CloudFormation project to associate with new resources." 

Also, is it possible to just generate this project name? Is this something the user would care about?

jeancochrane replied:
Agreed about your description, I like that phrasing better.

I think we probably could get away with generating the project name dynamically, but it's in here because `Project` and `Environment` seem to be standard variables for us in Terraform modules. If it seems like we're using it differently enough in this context that the reasoning behind those variables doesn't apply, I'd be comfortable removing it.

rbreslow replied:
If the goal of this project is to make Raster Vision accessible to those who don't have as much AWS knowledge, it seems like we could simplify things.

jeancochrane replied:
True, although the project also has the goal of being useful to Azavea RV users, which might suggest that we should keep it consistent with our other modules. Opened up https://github.com/azavea/raster-vision-aws/issues/17 to track discussion, since I think the decision is bigger than this PR.

rbreslow replied:
Would `Project` and `Environment` serve any purpose other than a tag? I can't see an area where we'd need a custom project name that the prefix can't provide. Also, I don't think this tool is supposed to be used outside of R&D efforts, so there wouldn't be a need to define any sort of environment, as this isn't going to be an application we're supporting.

jeancochrane replied:
`Project` is used in tags and in the security group description, and `Environment` is used to namespace resources. But I think your point stands, they don't seem to be very useful in this context. @hectcastro do you think it's safe to remove `Project` and `Environment` here? Just want to be very sure that I'm not going to make a change that messes with R&D's workflow.

hectcastro replied:
I think that the value of `Environment` goes down a bit within the context of how most people (internal or external) would use this template.

The value of `Prefix` is pretty clear (namespacing resources, for example, with `Prefix: Hector`). I also see value in `Project` because I might be working on two different projects under the same `Prefix` (e.g., `Project: Acme`, `Project: E Corp`).

I think you could overload `Prefix` to solve for the multi project case, but it seems more straightforward to have them separate.

jeancochrane replied:
An interesting snag here is that since `Project` is only used in tags and descriptions, `Prefix` ultimately controls the namespacing of resources -- meaning that you couldn't have two `Projects` under the same `Prefix` because the names of the AWS resources would be identical. (My understanding is that at least some Batch resources need unique names -- I know this is true for [job definitions](https://docs.aws.amazon.com/batch/latest/userguide/create-job-definition.html), for example -- but I haven't confirmed it for all resources in this template). We can't rely on using the value of `Project` for namespacing because CloudFormation doesn't support string functions and it doesn't make much sense to enforce slugified project names. So in the situation you describe, I think you'd have to have separate prefixes for both the `Acme` and the `E Corp` project.

hectcastro replied:
In that case, I think overloading `Prefix` (and dropping the others) seems like most straightforward approach.

jeancochrane replied:
Awesome, removed `Project` and `Environment`. Thanks all for sticking with this discussion and making strong cases, I think this is the right move now.

rbreslow commented on cloudformation/template.yml:
```diff
@@ -75,12 +71,25 @@ Parameters:
+    AllowedPattern: ^[a-zA-Z0-9]*$
+    ConstraintDescription: must only contain uppercase and lowercase letters and numbers
+
+  Slug:
```
I would replace usage of the word "Slug" with "Prefix." I think that better represents what the variable is used for.

jeancochrane replied:
Nice, that's much clearer.

rbreslow reviewed (approved):
Looks good. 👍

I'm not so familiar with traps you'd encounter in CloudFormation, so it may be worth having someone else take a second pass.

Commit:
Change 'Slug' parameter to 'Prefix'

Change the display name and variable name of the 'Slug' parameter
to 'Prefix' to more clearly indicate what it's for.

Commit:
Remove Project and Environment variables

The Project and Environment variables weren't accomplishing much
in the CloudFormation template, mostly used in tags. Remove them
to simplify the template.
